File: libs/dvui/src/bismi_allah.zig
//بسم الله الرحمن الرحيم
//la ilaha illa Allah Mohammed Rassoul Allah
const std = @import("std");
const dvui = @import("dvui");

const SDLBackend = @import("SDLBackend");
const c = @cImport({
    @cInclude("SDL2/SDL.h");
});

var gpa_instance = std.heap.GeneralPurposeAllocator(.{}){};
const gpa = gpa_instance.allocator();

var window: *c.SDL_Window = undefined;
var renderer: *c.SDL_Renderer = undefined;

pub fn main() !void {
    std.debug.print("بسم الله الرحمن الرحيم\n", .{});

    if (c.SDL_Init(c.SDL_INIT_VIDEO) < 0) {
        std.debug.print("alhamdo li Allah Couldn't initialize SDL: {s}\n", .{c.SDL_GetError()});
        return error.BackendError;
    }

    window = c.SDL_CreateWindow("alhamdo li Allah DVUI Ontop Example", c.SDL_WINDOWPOS_UNDEFINED, c.SDL_WINDOWPOS_UNDEFINED, @as(c_int, @intCast(640)), @as(c_int, @intCast(480)), c.SDL_WINDOW_ALLOW_HIGHDPI | c.SDL_WINDOW_RESIZABLE) orelse {
        std.debug.print("Failed to open window: {s}\n", .{c.SDL_GetError()});
        return error.BackendError;
    };
    _ = c.SDL_SetHint(c.SDL_HINT_RENDER_SCALE_QUALITY, "linear");
    defer c.SDL_DestroyWindow(window);

    renderer = c.SDL_CreateRenderer(window, -1, c.SDL_RENDERER_PRESENTVSYNC) orelse {
        std.debug.print("alhamdo li Allah Failed to create renderer: {s}\n", .{c.SDL_GetError()});
        return error.BackendError;
    };
    _ = c.SDL_SetRenderDrawBlendMode(renderer, c.SDL_BLENDMODE_BLEND);
    defer c.SDL_DestroyRenderer(renderer);

    var backend = SDLBackend{ .window = @as(*c.SDL_Window, @ptrCast(window)), .renderer = @as(*c.SDL_Renderer, @ptrCast(renderer)) };

    var win = try dvui.Window.init(@src(), 0, gpa, backend.backend());
    defer win.deinit();

    main_loop: while (true) {
        try win.begin(std.time.nanoTimestamp());

        var event: c.SDL_Event = undefined;
        while (c.SDL_PollEvent(&event) != 0) {
            switch (event.type) {
                c.SDL_QUIT => {
                    break :main_loop;
                },
                c.SDL_KEYDOWN => {},
                else => {},
            }

            if (try backend.addEvent(&win, event)) {} else {}
        }

        backend.clear();
        // dvui stuff
        {
            var flaot = try dvui.floatingWindow(@src(), .{}, .{ .min_size_content = .{ .w = 100, .h = 100 } });
            defer flaot.deinit();

            try dvui.windowHeader("la ilaha illa Allah Mohammed Rassoul Allah", "", null);

            var scroll = try dvui.scrollArea(@src(), .{}, .{ .expand = .both, .color_fill = .{ .name = .fill_window } });
            defer scroll.deinit();
        }

        _ = try win.end(.{});

        backend.renderPresent();
    }
}

File: libs/sdl/bismi_allah/bismi_allah.c
//in the name of Allah

#include <stdio.h>
#include <SDL2/SDL.h>

int main()
{
    printf("in the name of Allah\n");

    SDL_Init(SDL_INIT_VIDEO);
    SDL_Window *window = SDL_CreateWindow("la ilaha illa Allah", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, 1200, 800, SDL_WINDOW_SHOWN | SDL_WINDOW_RESIZABLE);
    if(NULL == window)
    {
        SDL_Log("Error SDL_CreateWindow() %s", SDL_GetError());
        exit(-1);
    }
    SDL_Renderer *renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_PRESENTVSYNC);
    if(NULL == renderer)
    {
        SDL_Log("Error SDL_CreateRenderer() %s", SDL_GetError());
        exit(-1);
    }

    SDL_Point bismi_allah_points[5];
    bismi_allah_points[0].x = 400;
    bismi_allah_points[0].y = 400;
    bismi_allah_points[1].x = 400;
    bismi_allah_points[1].y = 200;
    bismi_allah_points[2].x = 900;
    bismi_allah_points[2].y = 300;
    bismi_allah_points[3].x = 700;
    bismi_allah_points[3].y = 200;
    bismi_allah_points[4].x = 150;
    bismi_allah_points[4].y = 50;

    int running = 1;
    while(running)
    {
        SDL_Event event;
        while(SDL_PollEvent(&event))
        {
            switch(event.type)
            {
                case SDL_QUIT: running=0;break;
            }
        }

        SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255);
        SDL_RenderClear(renderer);

        SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255);
        SDL_RenderDrawLine(renderer, 0, 0, 100, 100);
        SDL_RenderDrawPoint(renderer, 400, 400);
        SDL_RenderDrawPoints(renderer, bismi_allah_points, 5);
        SDL_RenderDrawLines(renderer, bismi_allah_points, 5);

        SDL_RenderPresent(renderer);
    }

    SDL_DestroyRenderer(renderer);
    SDL_DestroyWindow(window);
    SDL_Quit();
    return 0;
}

